python/lsst/daf/butler/gen2convert/extractor.py

```diff
@@ -0,0 +1,159 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import re
+from collections import OrderedDict
+
+from .structures import Gen2Dataset, Gen2DatasetType
+
+__all__ = ("Extractor",)
+
+
+TEMPLATE_RE = re.compile(r'\%\((?P<name>\w+)\).*?(?P<type>[idrs])')
+
+
+class FilePathParser:
+    """A callable object that extracts Gen2Dataset instances from filenames
+    corresponding to a particular Gen2 DatasetType.
+
+    External code should use the `fromMapping` method to construct instances.
+
+    Parameters
+    ----------
+    datasetType : `Gen2DatasetType`
+        Information about the DatasetType this parser processes.
+    regex : regular expression object
+        Regular expression pattern with named groups for all data ID keys.
+    """
+
+    @classmethod
+    def fromMapping(cls, mapping):
+        """Construct a FilePathParser instance from a Gen2
+        `lsst.obs.base.Mapping` instance.
+        """
+        try:
+            template = mapping.template
+        except RuntimeError:
+            return None
+        datasetType = Gen2DatasetType(name=mapping.datasetType,
+                                      keys={},
+                                      persistable=mapping.persistable,
+                                      python=mapping.python)
+        iterator = TEMPLATE_RE.finditer(template)
+        try:
+            m1 = iterator.__next__()
+        except StopIteration:
+            regex = re.compile(re.escape(template))
+            return cls(datasetType=datasetType, regex=regex)
+        terms = [template[:m1.start()]]
+        allKeys = mapping.keys()
+        try:
+            while True:
+                name = m1.group("name")
+                if name == "patch":
+                    pattern = r"\d+,\d+"
+                elif m1.group("type") in "id":  # integers
+                    pattern = r"0*\d+"
+                else:
+                    pattern = ".+"
+                # only use named groups for the first occurence of a key
+                if name not in datasetType.keys:
+                    terms.append(r"(?P<%s>%s)" % (name, pattern))
+                    datasetType.keys[name] = allKeys[name]
+                else:
+                    terms.append(r"(%s)" % pattern)
+                m2 = iterator.__next__()
+                terms.append(re.escape(template[m1.end():m2.start()]))
+                m1 = m2
+        except StopIteration:
+            pass
+        terms.append(re.escape(template[m1.end():]))
+        return cls(datasetType=datasetType, regex=re.compile("".join(terms)))
```

Comment:
I had previously convinced myself that I couldn't use a for loop here because I needed to have both the current and next element available together, but of course that's silly.  Thanks for the extra push to simplify it; it's much better now.